core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java

```diff
@@ -359,21 +359,31 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
          return invokeNext(ctx, command);
       }
       CompletableFuture<?> l1InvalidationFuture = invalidateL1InCluster(ctx, command, assumeOriginKeptEntryInL1);
-      return invokeNextThenApply(ctx, command, (rCtx, rCommand, rv) -> {
+      return invokeNextAndHandle(ctx, command, (InvocationContext rCtx, VisitableCommand rCommand, Object rv, Throwable ex) -> {
          DataWriteCommand dataWriteCommand = (DataWriteCommand) rCommand;
-         if (l1InvalidationFuture != null && !l1InvalidationFuture.isDone() && isSynchronous(dataWriteCommand)) {
-            if (shouldRemoveFromLocalL1(rCtx, dataWriteCommand)) {
-               VisitableCommand removeFromL1Command = removeFromL1Command(rCtx, dataWriteCommand.getKey());
-               makeStage(asyncInvokeNext(rCtx, removeFromL1Command, l1InvalidationFuture))
-                     .thenApply(null, null, (rCtx2, rCommand2, rv2) -> rv);
+         if (ex != null) {
+            if (l1InvalidationFuture != null && !l1InvalidationFuture.isDone() && isSynchronous(dataWriteCommand) ) {
+               return asyncValue(l1InvalidationFuture).thenApply(rCtx, rCommand, (rCtx1, rCommand1, rv1) -> {
+                  throw ex;
+               });
             } else {
-               return asyncValue(l1InvalidationFuture.thenApply(nil -> rv));
+               throw ex;
+            }
+         } else {
+            if (l1InvalidationFuture != null && !l1InvalidationFuture.isDone() && isSynchronous(dataWriteCommand)) {
+               if (shouldRemoveFromLocalL1(rCtx, dataWriteCommand)) {
+                  VisitableCommand removeFromL1Command = removeFromL1Command(rCtx, dataWriteCommand.getKey());
+                  makeStage(asyncInvokeNext(rCtx, removeFromL1Command, l1InvalidationFuture))
```

Comment:
Missing `return` statement. Yes, the original code is missing that, too - that means that we haven't been waiting for the invalidation future to complete in the past either...